### PR TITLE
Diabetes linear regression example as e2e test

### PR DESCRIPTION
**Pull request by gordoncaleb**
This adds a linear regression example used in: 
http://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html

The data for this example comes from:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.rwrite1.txt

This PR also adds the ability to read in csv directly to a data array or one of the keanu tensor types:

Given a data class definition:
```
public static class Data {
     public DoubleTensor bmi;
     public DoubleTensor y;
 }
```
And some CSV data:

```
bmi                 ,y               
0.0380759064334241  ,0.0506801187398187
-0.00188201652779104,-0.044641636506989
0.0852989062966783  ,0.0506801187398187
-0.0890629393522603 ,-0.044641636506989
```

Read it in and create your Data object (containing tensors):
```
Data data = ReadCsv
    .fromResources("...myData.csv")
    .asVectorizedColumnsDefinedBy(Data.class)
    .load(true);
```

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java [hunk 1]
```diff
@@ -95,12 +95,60 @@ public void linearRegressionTwoFactorTensorVariationalMAP() {
...
+        DoubleVertex yGenerator = new GaussianVertex(yGeneratorMu, 1.0);
+        DoubleTensor yData = yGenerator.sample(random);
+
+        // Run Linear Regression
+        DoubleVertex[] weights = new DoubleVertex[featureCount];
+        DoubleVertex[] x = new DoubleVertex[featureCount];
+        DoubleVertex yMu = ConstantVertex.of(0.0);
+        for (int i = 0; i < weights.length; i++) {
+            weights[i] = new GaussianVertex(0.0, 1.0);
+            weights[i].setValue(0);
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java]: Do we need to do this?

File: keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java [hunk 2]
```diff
@@ -95,12 +95,60 @@ public void linearRegressionTwoFactorTensorVariationalMAP() {
...
+        optimizer.maxLikelihood();
 
         log.info("W1 = " + w1.getValue().scalar() + " W2 = " + w2.getValue().scalar() + ", B = " + b.getValue().scalar());
         assertEquals(expectedW1, w1.getValue().scalar(), 0.05);
         assertEquals(expectedW2, w2.getValue().scalar(), 0.05);
         assertEquals(expectedB, b.getValue().scalar(), 0.05);
     }
 
+    @Test
+    public void linearRegressionManyFactorTensorVariationalMAP() {
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java]: Cool test btw!

File: keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/bycolumn/CsvReaderByColumnTest.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+import io.improbable.keanu.util.csv.ReadCsv;
+import io.improbable.keanu.util.csv.pojo.CsvProperty;
+import org.junit.Test;
+
+import static org.apache.commons.lang3.ArrayUtils.toPrimitive;
+import static org.junit.Assert.assertArrayEquals;
+
+public class CsvReaderByColumnTest {
+
+    String csv = "a,b,c\n0,2.5,t\n1,3,true\n-3,4e3,0";
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/bycolumn/CsvReaderByColumnTest.java]: nit: can we format this to look like a preview of the csv file?

```
    String csv = "a,b,c\n" +
                 "0,2.5,t\n" +
                 "1,3,true\n" +
                 "-3,4e3,0";
```

File: keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/byrow/CsvReaderByRowTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import io.improbable.keanu.util.csv.pojo.CsvProperty;
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class CsvReaderByRowTest {
+
+    String csv = "a,b,c\nhel,lo,world";
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/byrow/CsvReaderByRowTest.java]: same nit as above:

```
    String csv = "a,b,c\n" +
                 "hel,lo,world";
```

File: keanu-project/src/test/java/io/improbable/keanu/e2e/regression/DiabetesLinearRegression.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        Data data = ReadCsv
+            .fromResources("data/datasets/diabetes/diabetes_standardized_training.csv")
+            .asVectorizedColumnsDefinedBy(Data.class)
+            .load(true);
+
+        // Linear Regression
+        DoubleVertex weight = new GaussianVertex(0.0, 2.0);
+        DoubleVertex b = new GaussianVertex(0.0, 2.0);
+        DoubleVertex x = ConstantVertex.of(data.bmi);
+        DoubleVertex yMu = x.multiply(weight);
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/e2e/regression/DiabetesLinearRegression.java]: It's the mu for the y variable, which is x*w.

File: keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java (hunk 1 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java]: Not really. Just added to make sure the optimizer was starting off in a fairly far away place.

File: keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/bycolumn/CsvReaderByColumnTest.java (hunk 3 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/bycolumn/CsvReaderByColumnTest.java]: Good call.

File: keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/byrow/CsvReaderByRowTest.java (hunk 4 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/util/csv/pojo/byrow/CsvReaderByRowTest.java]: Yup.

File: keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java [hunk 6]
```diff
@@ -114,11 +114,45 @@ public DoubleTensor max(DoubleTensor max) {
...
+        return tensor.sumNumber().doubleValue() / tensor.length();
+    }
+
+    @Override
+    public double standardDeviation() {
+        double average = average();
+        return Math.sqrt(Transforms.pow(tensor.sub(average), 2, false)
+            .sumNumber().doubleValue() / (tensor.length() - 1));
+    }
+
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java]: Nice

**Pull request merged**